lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -480,73 +545,24 @@ ESIConnectServicesPass::surfaceReqs(hw::HWMutableModuleLike mod,
-    SmallVector<Value, 16> newOperands(inst->getOperands().begin(),
-                                       inst->getOperands().end());
-    SmallVector<Type, 16> newResultTypes(inst->getResultTypes().begin(),
-                                         inst->getResultTypes().end());
-
-    // Add new inputs for the new to_client requests and clone the request
-    // into the module containing `inst`.
-    circt::BackedgeBuilder beb(b, mod.getLoc());
-    SmallVector<circt::Backedge, 8> newResultBackedges;
+    // Add new inputs for the new bundles being requested.
+    SmallVector<Value, 16> newOperands;
     for (auto req : reqs) {
-      auto clone = cast<ServiceReqOpInterface>(b.clone(*req));
-      clone.setClientNamePath(
-          prependNamePart(clone.getClientNamePath(), inst.getInstanceName()));
-      if (Value toClient = clone.getToClient())
-        newOperands.push_back(toClient);
-      if (Type toServerType = clone.getToServerType()) {
-        newResultTypes.push_back(toServerType);
-        Backedge result = beb.get(toServerType);
-        newResultBackedges.push_back(result);
-        clone.setToServer(result);
-      }
+      auto clone = b.create<ServiceImplementConnReqOp>(
+          req.getLoc(), req.getToClient().getType(), req.getServicePortAttr(),
+          prependNamePart(req.getClientNamePath(), inst.getInstanceName()));
+      clone->setDialectAttrs(req->getDialectAttrs());
+      newOperands.push_back(clone.getToClient());
     }
-
-    // Create a replacement instance of the same operation type.
-    SmallVector<NamedAttribute> newAttrs;
-    for (auto attr : inst->getAttrs()) {
-      if (attr.getName() == argsAttrName) {
-        auto names = mod.getInputNames();
-        newAttrs.push_back(b.getNamedAttr(argsAttrName, b.getArrayAttr(names)));
-      } else if (attr.getName() == resultsAttrName) {
-        auto names = mod.getOutputNames();
-        newAttrs.push_back(
-            b.getNamedAttr(resultsAttrName, b.getArrayAttr(names)));
-      } else {
-        newAttrs.push_back(attr);
-      }
-    }
-    auto *newInst = b.insert(Operation::create(
-        inst->getLoc(), inst->getName(), newResultTypes, newOperands,
-        b.getDictionaryAttr(newAttrs), inst->getPropertiesStorage(),
-        inst->getSuccessors(), inst->getRegions()));
-    newModuleInstantiations.push_back(
-        cast<igraph::InstanceOpInterface>(newInst));
-
-    // Replace all uses of the instance being replaced.
-    for (auto [newV, oldV] :
-         llvm::zip(newInst->getResults(), inst->getResults()))
-      oldV.replaceAllUsesWith(newV);
-
-    // Clone the to_server requests and wire them up to the new instance.
-    outputCounter = origNumOutputs;
-    for (Backedge newResult : newResultBackedges)
-      newResult.setValue(newInst->getResult(outputCounter++));
+    inst->insertOperands(inst->getNumOperands(), newOperands);
+    // Set the names, if we know how.
+    if (auto hwInst = dyn_cast<hw::InstanceOp>(*inst))
```

Comment:
I suggest opening an issue to track locations where we really should be using a `HWMutableInstanceLike` interface.

Reply:
#6319